Consolidate buffer packing functions with less atomics

alexrlongne opened the pull request:
## PR Summary

* Profiling with NSIGHT systems revealed that the `LoadBuffer_` function in `swarm_comms.cpp` took 20% of the GPU time in the particles example that used much more particles (1e6 per block). NSIGHT compute showed that the expensive kernel in `LoadBuffer_` had many stalls, we speculated this was due to waiting on the `atomic_fetch_add` function in `LoadBuffer_`. This PR reworks `LoadBuffer_` and `CountParticlesToSend_` to remove atomics and instead sort the particles by buffer ID and use their sorted order to determine their size and index into that buffer.

* Removing atomics reduces the runtime by about 20% in the million particle example and more in a 10 million particle version. A case where the particles are photons may involves even more cell block crossings as c*delta_t could be large compared to a block size (in the particles example the particle's path length is equal to a block size).  

* The reason for the atomic in both cases was that the loops were over particles where the particle buffer was determined and that was then used to atomically update a buffer specific field.

* As a rider, this PR also updates some sampling functions in the particle example to use less transcendentals, which has shown improvement in a Parthenon downstream code. I can separate into another PR if desired.

## Changes
+ Fold in buffer sizing to the load buffer function
+ Use a sort and a discontinuity check instead of an atomic_fetch_add inside of LoadBuffer_ to get particle index into buffer
+ Reduce transcendental functions call in particle sourcing in particle example

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [ ] Code passes cpplint
- [ ] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [ ] Code is formatted
- [ ] Changes are summarized in CHANGELOG.md
- [ ] Change is breaking (API, behavior, ...)
  - [ ] Change is *additionally* added to CHANGELOG.md in the breaking section
  - [ ] PR is marked as breaking
  - [ ] Short summary API changes at the top of the PR (plus optionally with an automated update/fix script)
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [ ] Docs build
- [ ] (@lanl.gov employees) Update copyright on changed files

alexrlongne committed:
Consolidate buffer packing functions with less atomics

+ Fold in buffer sizing to the load buffer function
+ Use a sort and a discontinuity check instead of an
atomic_fetch_add inside of LoadBuffer_ to get particle
index into buffer
+ Reduce transcendental functions call in particle sourcing in
particle example

alexrlongne commented on src/interface/swarm_comms.cpp:
```diff
@@ -280,7 +344,7 @@ void Swarm::Send(BoundaryCommSubset phase) {
 
   // Query particles for those to be sent
-  CountParticlesToSend_();
+  //CountParticlesToSend_();
```
I'll remove this commented out function call and remove the old function.

alexrlongne replied:
Done.

alexrlongne commented on src/interface/swarm_comms.cpp:
```diff
@@ -236,42 +240,102 @@ void Swarm::LoadBuffers_() {
-      KOKKOS_LAMBDA(const int n) { buffer_counters[n] = 0; });
+  if(max_active_index_ >= 0) {
+    // Make an n particle sized array of index, buffer pairs (with SwarmKey struct)
+    ParArray1D<SwarmKey> buffer_sorted("buffer_sorted", max_active_index_+1);
```
This PR reuses the `SwarmKey` struct that used to sort and count particles in a cell. I can make a new version of that struct with more appropriate member names if you'd like.

brryan replied:
What about renaming `cell_idx_1d_` in `SwarmKey` to something generic like `sort_idx_`? Then that one struct could be sensibly shared with the cell sorting and this buffer sorting.

I also think we want to get this `ParArray1D` declaration out of the `LoadBuffers_` call, because that will be a dynamic memory allocation every cycle which can be pretty slow. I would mimic what we do with the persistent memory for cell sorts, by adding `ParArray1D<SwarmKey> buffer_sorted_` right below `ParArray1D<SwarmKey> cell_sorted_` in `swarm.hpp`, and then also duplicating the work of initially allocating and resizing `buffer_sorted_` like we do `cell_sorted_`.

brryan replied:
Ditto for the `buffer_start` just below this comment (although it won't need to be resized as the swarm pool increases)

alexrlongne replied:
I've added these to the class data and changed the `cell_idx_1d_` name to `sort_idx_`. Thanks!

brryan commented on src/interface/swarm_comms.cpp:
```diff
@@ -236,42 +240,102 @@ void Swarm::LoadBuffers_() {
+            buffer_sorted(n) = SwarmKey(m, n);
+          }
+          else {
+            buffer_sorted(n) = SwarmKey(-1, n);
```
```suggestion
            buffer_sorted(n) = SwarmKey(this_block_, n);
```
This is a `Swarm` member we use to store the `-1` magic number as much as possible

alexrlongne replied:
Fixed, thanks!

brryan commented on src/interface/swarm_comms.cpp:
```diff
@@ -236,42 +240,102 @@ void Swarm::LoadBuffers_() {
+
+          // end checks (used to to size particle buffers)
+          if (m >= 0 && n == max_active_index ) {
+            num_particles_to_send(m) = n +1;
```
@alexrlongne in parthenon we use `clang-format-12` for formatting (I just have this hooked up to vscode so I always format on save) but just mentioning because the CI will fail otherwise.

alexrlongne replied:
Thanks, the pipeline's passing for that now so I'll mark this as resolved.

brryan commented on src/interface/swarm_comms.cpp:
```diff
@@ -236,42 +240,102 @@ void Swarm::LoadBuffers_() {
+            num_particles_to_send(m) = n +1;
+          }
+          else if (m >= 0 && m != buffer_sorted(n+1).cell_idx_1d_ ) {
+            num_particles_to_send(m) = n +1;
```
This is slick

alexrlongne replied:
Thanks! Glad this worked out with minimal overhead.

brryan reviewed (commented):
@alexrlongne This looks great, I think with just formatting, cleanup of the `SwarmKey` `cell_idx_1d` name, and removing the dynamic memory allocs, this is ready to go.

alexrlongne committed:
Address PR comments

+ Call the member variable in SwarmKey the sort_key
+ Remove CountParticlesInBuffer function
+ Add buffer_start and buffer_sorted as swarm member variables

brryan commented on example/particles/particles.cpp:
```diff
@@ -288,17 +288,18 @@ TaskStatus CreateSomeParticles(MeshBlock *pmb, const double t0) {
-          v(0, n) = v(0, n) - vdN / NdN * x(n);
-          v(1, n) = v(1, n) - vdN / NdN * y(n);
-          v(2, n) = v(2, n) - vdN / NdN * z(n);
+          Real inverse_NdN = r * r;
```
```suggestion
          Real inverse_NdN = 1. / (r * r);
```

alexrlongne replied:
Whoops sorry! I guess the orbiting particles option is not in the tests right now. Applying...

brryan commented on src/interface/swarm_comms.cpp:
```diff
@@ -156,72 +156,15 @@ void Swarm::SetupPersistentMPI() {
+  vbswarm->particle_size = particle_size;
   const int nneighbor = pmb->neighbors.size();
+  // Fence to make sure particles aren't currently being transported locally
+  // TODO(BRR) do this operation on device.
```
```suggestion
```
Pretty sure that's a leftover comment I should have removed in an earlier cleanup.

brryan reviewed (approved):
Looks great! Just one more query about r^2 vs r^-2

alexrlongne committed:
Update example/particles/particles.cpp

Co-authored-by: Ben Ryan <bryan10@illinois.edu>

alexrlongne committed:
Update src/interface/swarm_comms.cpp

Co-authored-by: Ben Ryan <bryan10@illinois.edu>

brryan committed:
Merge branch 'develop' into along/try_less_atomic
